ion-schema/src/isl/version_based_isl/mod.rs

```diff
@@ -0,0 +1,396 @@
+        name: A,
+        constraints: B,
+    ) -> IslType<V> {
+        IslType::new(Some(name.into()), constraints.into())
+    }
+
+    /// Creates an anonymous [IslType] using the [IslConstraint] defined within it
+    pub fn anonymous<A: Into<Vec<IslConstraint<V>>>>(constraints: A) -> IslType<V> {
+        IslType::new(None, constraints.into())
+    }
+
+    /// Provides a name if the ISL type is named type definition
+    /// Otherwise returns None
+    pub fn name(&self) -> &Option<String> {
+        &self.name
+    }
+
+    /// Provides open content that is there in the type definition
+    pub fn open_content(&self) -> Vec<(&String, &Element)> {
+        let mut open_contents = vec![];
+        for constraint in &self.constraints {
+            if let IslConstraint::Unknown(open_content) = constraint {
+                open_contents.push((open_content.field_name(), open_content.field_value()))
+            }
+        }
+        open_contents
+    }
+
+    /// Provides the underlying constraints of [IslType]
+    pub fn constraints(&self) -> &[IslConstraint<V>] {
+        &self.constraints
+    }
+}
+
+/// Provides an internal representation of a schema type reference.
+/// The type reference grammar is defined in the [Ion Schema Spec]
+///
+/// [Ion Schema spec]: https://amazon-ion.github.io/ion-schema/docs/isl-1-0/spec#grammar
+#[derive(Debug, Clone, PartialEq)]
+pub enum IslTypeRef<V: IslVersionTrait> {
+    /// Represents a reference to a named type (including aliases and built-in types)
+    Named(String, NullabilityModifier),
+    /// Represents a type reference defined as an inlined import of a type from another schema
+    TypeImport(IslImportType, NullabilityModifier),
+    /// represents an unnamed type definition reference
+    Anonymous(IslType<V>, NullabilityModifier),
+}
+
+impl<V: IslVersionTrait> IslTypeRef<V> {
+    /// Creates a named [IslTypeRef] using the name of the type referenced inside it
+    pub fn named<A: Into<String>>(name: A) -> IslTypeRef<V> {
+        IslTypeRef::Named(name.into(), NullabilityModifier::Nothing)
+    }
+
+    /// Creates a nullable [IslTypeRef] using the [IonType] referenced inside it
+    pub fn nullable_built_in(name: IonType) -> IslTypeRef<IslV1_0> {
+        IslTypeRef::Named(format!("{name}"), NullabilityModifier::Nullable)
+    }
+
+    /// Creates an anonymous [IslTypeRef] using the [IslConstraint]s referenced inside it
+    pub fn anonymous<A: Into<Vec<IslConstraint<V>>>>(constraints: A) -> IslTypeRef<V> {
+        IslTypeRef::Anonymous(
+            IslType::new(None, constraints.into()),
+            NullabilityModifier::Nothing,
+        )
+    }
+
+    /// Creates an [IslVariablyOccurringTypeRef] using the [IslConstraint]s and [Range] referenced inside it
+    pub fn variably_occurring(
+        type_ref: IslTypeRef<V>,
+        occurs: Range,
+    ) -> IslVariablyOccurringTypeRef<V> {
+        IslVariablyOccurringTypeRef::new(type_ref, occurs)
+    }
```

Comment:
I don't think this belongs here because it is a constructor for `IslVariableOccuringTypeRef`.

Reply:
Changed this to be part of `IslVariableOccuringTypeRef`